Source/CombatExtended/Compatibility/TurretRegistry.cs

```diff
@@ -0,0 +1,185 @@
+	private static Dictionary<Type, GetReloadingFunction> getReloading;
+	private static Dictionary<Type, GetAmmoFunction> getAmmo;
+	private static Dictionary<Type, GetGunFunction> getGun;
+	
+        private static void Enable()
+        {
+            enabled = true;
+	    setReloading = new Dictionary<Type, SetReloadingFunction>();
+	    getReloading = new Dictionary<Type, GetReloadingFunction>();
+	    getAmmo      = new Dictionary<Type, GetAmmoFunction>();
+	    getGun       = new Dictionary<Type, GetGunFunction>();
+	    
+        }
+
+        public static void RegisterReloadableTurret(Type turretType, SetReloadingFunction setReload, GetReloadingFunction getReload, GetGunFunction gun, GetAmmoFunction ammo=null)
+        {
+            if (!enabled)
+            {
+                Enable();
+            }
+	    getReloading[turretType] = getReload;
+	    setReloading[turretType] = setReload;
+
+	    getAmmo[turretType] = ammo;
+	    getGun[turretType] = gun;
+        }
+
+	public static void SetReloading(this Building_Turret turret, bool reloading) {
+	    var ceturret = turret as Building_TurretGunCE;
+	    if (ceturret!=null) {
+		ceturret.isReloading = reloading;
+		return;
+	    }
+	    if (enabled) {
+		SetReloadingFunction func;
+		if (setReloading.TryGetValue(turret.GetType(), out func)) {
+		    func(turret, reloading);
+		    return;
+		}
+	    }
+	    CELogger.Warn("Asked to set reloading on an unknown turret type: "+turret);
+	}
+
+	public static bool GetReloading(this Building_Turret turret) {
+	    var ceturret = turret as Building_TurretGunCE;
+	    if (ceturret!=null) {
+		return ceturret.isReloading;
+	    }
+	    if (enabled) {
+		GetReloadingFunction func;
+		if (getReloading.TryGetValue(turret.GetType(), out func)) {
+		    return func(turret);
+		}
+	    }
+	    CELogger.Warn("Asked to get reloading on an unknown turret type: "+turret);
+	    return false;
+	}
+
+	public static CompAmmoUser GetAmmo(this Building_Turret turret) {
+	    var ceturret = turret as Building_TurretGunCE;
+	    if (ceturret!=null) {
+		return ceturret.CompAmmo;
+	    }
+	    if (enabled) {
+		GetAmmoFunction func;
+		if (getAmmo.TryGetValue(turret.GetType(), out func)) {
+		    return func(turret);
+		}
+		GetGunFunction gfunc;
+		if (getGun.TryGetValue(turret.GetType(), out gfunc)) {
+		    return gfunc(turret)?.TryGetComp<CompAmmoUser>();
+		}
+		
+	    }
+	    CELogger.Warn("Asked to get ammo on an unknown turret type: "+turret);
+	    return null;
+	}
+
+	public static Thing GetGun(this Building_Turret turret) {
+	    var ceturret = turret as Building_TurretGunCE;
+	    if (ceturret!=null) {
+		return ceturret.Gun;
+	    }
+	    if (enabled) {
+		GetGunFunction gfunc;
+		if (getGun.TryGetValue(turret.GetType(), out gfunc)) {
+		    return gfunc(turret);
+		}
+		
+	    }
+	    CELogger.Warn("Asked to get gun on an unknown turret type: "+turret);
+	    return null;
+	}
+
+	public static CompMannable GetMannable(this Building_Turret turret) {
```

Comment:
```csharp
public static CompMannable GetMannable(this Building_Turret turret) {
            if(turret is Building_TurretGunCE ceturret) return ceturret.MannableComp;
	    return turret.TryGetComp<CompMannable>();
}
```

Reply:
At the time I first wrote this, I was trying to keep to mono-compiler C#7-compatible language features.  There's probably a few other places that can see a similar benefit scattered throughout the code base.